[move-lang] Attributes

tnowacki opened the pull request:
- Added attributes to Move
 - Syntax is based on Rust attribute syntax, which is in turn based on the standards found in ECMA-334 and ECMA-335
 - Attributes can currently be added on address blocks, modules, scripts, and any module top level member. 
   - Eventually they should be able to be parsed on expression blocks, expression spec blocks, and expression use statements. 
   
- There is no sanity checking currently 
  - we probably eventually want some way for attributes to register with the compiler so they can be checked
  - We will likely hard code this temporarily for unit tests

## Motivation

- Blocking for unit test framework
- Should be helpful for other tools too, particularly the prover. 

## Test Plan

- Added a few tests 

## Related PRs

Built on top of #8143, so waiting for that to land

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -392,6 +392,78 @@ fn parse_visibility(tokens: &mut Lexer) -> Result<Visibility, Error> {
+
+// Parse attributes. Used to annotate a variety of AST nodes
+//      Attributes = ("#[" Comma<Attribute> "]")*
+fn maybe_parse_attributes(tokens: &mut Lexer) -> Result<Vec<Attributes>, Error> {
```
This should be called `parse_attributes` to follow the naming convention -- it's parsing the "Attributes" non-terminal. The fact that "Attributes" may be empty doesn't mean we need a "maybe" prefix on the name.

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -392,6 +392,78 @@ fn parse_visibility(tokens: &mut Lexer) -> Result<Visibility, Error> {
+    while let Tok::AttrBegin = tokens.peek() {
+        let start_loc = tokens.start_loc();
+        tokens.advance()?;
+        let attributes_ = parse_comma_list_after_start(
```
I'm curious: is there a particular reason why you want to combine the "#" and "[" in a single AttrBegin token instead of using a single "#" token with a call here to `parse_comma_list` with an LBracket token for the `start_token`? I think it works either way, but maybe I'm missing something.

tnowacki replied:
No real reason, I can and will change this

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -1706,7 +1788,8 @@ fn parse_address_block(tokens: &mut Lexer) -> Result<(Loc, Address, Vec<ModuleDe
     let mut modules = vec![];
     while tokens.peek() != Tok::RBrace {
-        modules.push(parse_module(tokens)?);
+        let attributes = maybe_parse_attributes(tokens)?;
```
This doesn't match your change to the grammar comment. I'm not sure which one you want. The grammar shows an Attributes list before the "address", but the code parses an Attributes list before each module.

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -1895,11 +1996,13 @@ fn parse_script(tokens: &mut Lexer) -> Result<Script, Error> {
     let mut specs = vec![];
-    while tokens.peek() == Tok::Spec {
-        specs.push(parse_spec_block(tokens)?)
+    while tokens.peek() == Tok::AttrBegin || tokens.peek() == Tok::Spec {
```
Could you add this to the grammar comment? It was missing before, too.

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -2514,12 +2620,13 @@ fn parse_spec_property(tokens: &mut Lexer) -> Result<PragmaProperty, Error> {
 fn parse_file(tokens: &mut Lexer) -> Result<Vec<Definition>, Error> {
```
The grammar comment for File needs to be updated.

bob-wilson reviewed (commented):
I left some comments and a question. I'm also wondering if we really want to add support for attributes on all of these things instead of adding them more selectively and waiting to see if we need them in other places.

tnowacki commented:
@bob-wilson Thanks for the review! I'll be addressing the comments shortly. 

>  I'm also wondering if we really want to add support for attributes on all of these things instead of adding them more selectively and waiting to see if we need them in other places.

I added them to all elements where they will likely be needed for testing. Which is just about every top level item unfortunately...

tzakian commented on language/move-lang/tests/move_check/parser/attribute_placement.move:
```diff
@@ -0,0 +1,46 @@
+    public fun bar() {}
+}
+
+#[attr]
```
Realize this is a draft so probably more tests incoming, but just mentioning here since it would be nice to test:
* Multiple attributes (both same, and different)
* Parametrized attributes

tnowacki replied:
See attribute_variants.move, but I'll update that one to handle the same/different case

wrwg commented on language/move-lang/src/expansion/ast.rs:
```diff
@@ -123,6 +153,7 @@ pub struct Constant {
 
 #[derive(Debug, Clone, PartialEq)]
 pub struct SpecBlock_ {
+    pub attributes: Vec<Attribute>,
```
I wonder whether we can also have attributes on spec block members. This would likely capture all of the provers current use cases of our kind of attributes (we do have pragma on spec block level but also stuff like `ensures [attributes] ...`. We would probably just push this into each variant of SpecBlockMember.

Otherwise looks great to me.

tnowacki replied:
We can definitely add these on each member. I'm not so sure if you want to do things like `ensures #[attributes] ...`, but  I don't know enough about the syntax around that spot to know if it is the best place. 

Conceptually, attributes could be attached anywhere you would also allow comments. But for sanity and consistency, we will likely want to keep them at top level members, or for statement-like positions.

wrwg replied:
I was thinking of a syntax like

```
spec X {
  #[attribute]
  ensures P;
```

This should be more consistent.

BTW there is no hurry to do this in the first iteration, just wanted to bring this on your radar.

tnowacki replied:
That should definitely work!

tzakian reviewed (approved):
Looks good to me. Thank you for implementing this!

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -392,6 +392,77 @@ fn parse_visibility(tokens: &mut Lexer) -> Result<Visibility, Error> {
+        }
+        Tok::LParen => {
+            let start_loc = tokens.start_loc();
+            tokens.advance()?;
```
You can drop this `advance` and use `parse_comma_list` below.

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -392,6 +392,77 @@ fn parse_visibility(tokens: &mut Lexer) -> Result<Visibility, Error> {
+}
+
+// Parse attributes. Used to annotate a variety of AST nodes
+//      Attributes = ("#[" Comma<Attribute> "]")*
```
Nit: the NumSign and LBracket are separate tokens, so I suggest writing that as "#" "[" instead of "#["

bob-wilson commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -1706,7 +1787,8 @@ fn parse_address_block(tokens: &mut Lexer) -> Result<(Loc, Address, Vec<ModuleDe
     let mut modules = vec![];
     while tokens.peek() != Tok::RBrace {
-        modules.push(parse_module(tokens)?);
+        let attributes = parse_attributes(tokens)?;
```
The grammar for AddressBlock should have `( <Attributes> <Module> )*`

bob-wilson reviewed (approved):
I found a few more little things, but I'll go ahead and approve now.

tnowacki commented:
/land
